src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -240,35 +262,119 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
         };
         self.iterator.seek(full_key).await?;
         if !self.iterator.is_valid() {
+            self.state.invalidate();
             return Ok(());
         }
 
         if self.key_out_of_range() {
-            self.out_of_range = true;
+            self.state.invalidate();
             return Ok(());
         }
 
         self.delete_range_iter.seek(full_key.user_key).await?;
 
-        // Handle multi-version
-        self.last_key = FullKey::default();
-        // Handle range scan when key > end_key
-
-        self.next().await
+        self.step_one().await?;
+        if self.state.is_next() {
+            self.next().await?;
+        }
+        Ok(())
     }
 
     /// Indicates whether the iterator can be used.
     pub fn is_valid(&self) -> bool {
-        // Handle range scan
-        // key >= begin_key is guaranteed by seek/rewind function
-        (!self.out_of_range) && self.iterator.is_valid()
+        self.state.is_ready()
     }
 
     pub fn collect_local_statistic(&self, stats: &mut StoreLocalStatistic) {
         stats.add(&self.stats);
         self.iterator.collect_local_statistic(stats);
     }
 
+    /// Transition iterator state.
+    /// See `IteratorState` for the state machine details.
+    async fn step_one(&mut self) -> HummockResult<()> {
+        match self.state {
+            IteratorState::Init => {
+                self.last_key = FullKey::default();
+                self.on_next().await?;
```

Comment:
It looks correct to me. However, I did not find the pattern of calling `on_next`. Can add more comments to describe it

Reply:
I have refactored the logic:
- The original `step_one` and `next` are combined.
- `on_next` is renamed to `step_one` to make the logic easier to understand.